core/src/main/java/hudson/model/AbstractItem.java

```diff
@@ -801,6 +802,24 @@ public void delete() throws IOException, InterruptedException {
                     }
                 }
             }
+            if (this instanceof ItemGroup) {
+                // delete individual items first
+                // (disregard whether they would be deletable in isolation)
+                // JENKINS-34939: do not hold the monitor on this folder while deleting them
+                // (thus we cannot do this inside performDelete)
+                try (ACLContext oldContext = ACL.as2(ACL.SYSTEM2)) {
+                    for (Item i : ((ItemGroup<?>) this).getItems(TopLevelItem.class::isInstance)) {
```

Comment:
For now, retaining original behavior which is to say not calling `delete()` on children of non-top-level item groups (e.g., `MatrixConfiguration`). As in https://github.com/jenkinsci/cloudbees-folder-plugin/pull/356#discussion_r1372410882, the only practical difference is likely to be in whether `ItemListener` is notified of these.